### PR TITLE
Add documentation for CA1868

### PR DESCRIPTION
**Pull request by mpidash**
See issue https://github.com/dotnet/runtime/issues/85490 and PR https://github.com/dotnet/roslyn-analyzers/pull/6767 (not merged yet).

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/fundamentals/code-analysis/quality-rules/ca1868.md](https://github.com/dotnet/docs/blob/d5caccd09f72434741cb9a94cb3da5642ef585f4/docs/fundamentals/code-analysis/quality-rules/ca1868.md) | [docs/fundamentals/code-analysis/quality-rules/ca1868](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1868?branch=pr-en-us-36379) |
| [docs/fundamentals/code-analysis/quality-rules/index.md](https://github.com/dotnet/docs/blob/d5caccd09f72434741cb9a94cb3da5642ef585f4/docs/fundamentals/code-analysis/quality-rules/index.md) | [Code quality rules](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/index?branch=pr-en-us-36379) |
| [docs/fundamentals/code-analysis/quality-rules/performance-warnings.md](https://github.com/dotnet/docs/blob/d5caccd09f72434741cb9a94cb3da5642ef585f4/docs/fundamentals/code-analysis/quality-rules/performance-warnings.md) | [Performance rules](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/performance-warnings?branch=pr-en-us-36379) |

### CODE REVIEW

File: docs/fundamentals/code-analysis/quality-rules/ca1865.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+| **Fix is breaking or non-breaking** | Non-breaking                           |
+| **Enabled by default in .NET 7**    | No                                     |
+
+## Cause
+
+An <xref:System.Collections.Generic.ISet%601.Add%2A> or <xref:System.Collections.Generic.ICollection%601.Remove%2A> (<xref:System.Collections.Immutable.IImmutableSet%601.Add%2A?displayProperty=nameWithType> or <xref:System.Collections.Immutable.IImmutableSet%601.Remove%2A?displayProperty=nameWithType>) call is guarded by a call to <xref:System.Collections.Generic.ICollection%601.Contains%2A> (<xref:System.Collections.Immutable.IImmutableSet%601.Contains%2A?displayProperty=nameWithType>).
+
+## Rule description
+
+Both <xref:System.Collections.Generic.ISet%601.Add%2A> and <xref:System.Collections.Generic.ICollection%601.Remove%2A> perform a lookup which makes calling <xref:System.Collections.Generic.ICollection%601.Contains%2A> before redundant. It is more efficient to call <xref:System.Collections.Generic.ISet%601.Add%2A> and <xref:System.Collections.Generic.ICollection%601.Remove%2A> directly, which will return a Boolean value indicating whether an item has been added or removed.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1865.md:32]:
```suggestion
Both <xref:System.Collections.Generic.ISet%601.Add(%600)?displayProperty=nameWithType> and <xref:System.Collections.Generic.ICollection%601.Remove(%600)?displayProperty=nameWithType> perform a lookup, which makes it redundant to call <xref:System.Collections.Generic.ICollection%601.Contains(%600)?displayProperty=nameWithType> beforehand. It's more efficient to call <xref:System.Collections.Generic.ISet%601.Add(%600)> or<xref:System.Collections.Generic.ICollection%601.Remove(%600)> directly, which returns a Boolean value indicating whether the item was added or removed.
```

File: docs/fundamentals/code-analysis/quality-rules/ca1865.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+This also applies to <xref:System.Collections.Immutable.IImmutableSet%601.Add%2A?displayProperty=nameWithType> and <xref:System.Collections.Immutable.IImmutableSet%601.Remove%2A?displayProperty=nameWithType>, except that they either return a new set when an item is added or removed, or the original set otherwise.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1865.md:34]:
```suggestion
This logic also applies to <xref:System.Collections.Immutable.IImmutableSet%601.Add(%600)?displayProperty=nameWithType> and <xref:System.Collections.Immutable.IImmutableSet%601.Remove(%600)?displayProperty=nameWithType>, except that they either return a new set if the item is added or removed, or the original set if it wasn't.
```

File: docs/fundamentals/code-analysis/quality-rules/ca1865.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+## How to fix violations
+
+Replace the call to <xref:System.Collections.Generic.ICollection%601.Contains%2A> (<xref:System.Collections.Immutable.IImmutableSet%601.Contains%2A?displayProperty=nameWithType>) that is followed by a call to <xref:System.Collections.Generic.ISet%601.Add%2A> or <xref:System.Collections.Generic.ICollection%601.Remove%2A> (<xref:System.Collections.Immutable.IImmutableSet%601.Add%2A?displayProperty=nameWithType> or <xref:System.Collections.Immutable.IImmutableSet%601.Remove%2A?displayProperty=nameWithType>) with a single call to the latter.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1865.md:38]:
```suggestion
Replace the call to <xref:System.Collections.Generic.ICollection%601.Contains(%600)?displayProperty=nameWithType> (or <xref:System.Collections.Immutable.IImmutableSet%601.Contains(%600)?displayProperty=nameWithType>) that's followed by a call to <xref:System.Collections.Generic.ISet%601.Add(%600)?displayProperty=nameWithType> or <xref:System.Collections.Generic.ICollection%601.Remove(%600)?displayProperty=nameWithType> (or <xref:System.Collections.Immutable.IImmutableSet%601.Add(%600)?displayProperty=nameWithType> or <xref:System.Collections.Immutable.IImmutableSet%601.Remove(%600)?displayProperty=nameWithType>) with a single call to the latter method.
```

File: docs/fundamentals/code-analysis/quality-rules/index.md [hunk 4]
```diff
@@ -158,6 +158,7 @@ The following table lists code quality analysis rules.
...
 > | [CA1864: Prefer the 'IDictionary.TryAdd(TKey, TValue)' method](ca1864.md) | Both <xref:System.Collections.Generic.Dictionary%602.ContainsKey(%600)?displayProperty=nameWithType> and <xref:System.Collections.Generic.Dictionary%602.Add%2A?displayProperty=nameWithType> perform a lookup, which is redundant. It's is more efficient to call <xref:System.Collections.Generic.Dictionary%602.TryAdd%2A?displayProperty=nameWithType>, which returns a `bool` indicating if the value was added or not. `TryAdd` doesn't overwrite the key's value if the key is already present. |
+> | [CA1865: Unnecessary call to 'Contains' for sets](ca1865.md) | Both <xref:System.Collections.Generic.ISet%601.Add%2A> and <xref:System.Collections.Generic.ICollection%601.Remove%2A> perform a lookup which makes calling <xref:System.Collections.Generic.ICollection%601.Contains%2A> before redundant. It is more efficient to call <xref:System.Collections.Generic.ISet%601.Add%2A> and <xref:System.Collections.Generic.ICollection%601.Remove%2A> directly, which will return a Boolean value indicating whether an item has been added or removed. |
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/index.md:161]: Please update to match suggestion from main file.

File: docs/fundamentals/code-analysis/quality-rules/performance-warnings.md [hunk 5]
```diff
@@ -71,3 +71,4 @@ Performance rules support high-performance libraries and applications.
...
 | [CA1864: Prefer the 'IDictionary.TryAdd(TKey, TValue)' method](ca1864.md) | Both <xref:System.Collections.Generic.Dictionary%602.ContainsKey(%600)?displayProperty=nameWithType> and <xref:System.Collections.Generic.Dictionary%602.Add%2A?displayProperty=nameWithType> perform a lookup, which is redundant. It's is more efficient to call <xref:System.Collections.Generic.Dictionary%602.TryAdd%2A?displayProperty=nameWithType>, which returns a `bool` indicating if the value was added or not. `TryAdd` doesn't overwrite the key's value if the key is already present. |
+| [CA1865: Unnecessary call to 'Contains' for sets](ca1865.md) | Both <xref:System.Collections.Generic.ISet%601.Add%2A> and <xref:System.Collections.Generic.ICollection%601.Remove%2A> perform a lookup which makes calling <xref:System.Collections.Generic.ICollection%601.Contains%2A> before redundant. It is more efficient to call <xref:System.Collections.Generic.ISet%601.Add%2A> and <xref:System.Collections.Generic.ICollection%601.Remove%2A> directly, which will return a Boolean value indicating whether an item has been added or removed. |
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/performance-warnings.md:74]: Please update to match suggestion from main file.

**Review by gewarren — COMMENTED** [review]
Thanks for adding the docs @mpidash!

File: docs/fundamentals/code-analysis/quality-rules/index.md (hunk 4 above)
Comment (mpidash) [reply docs/fundamentals/code-analysis/quality-rules/index.md]: Updated in [e055b1c](https://github.com/dotnet/docs/pull/36379/commits/e055b1cae46fe07e6c02527aff246dce1e4ada9c).

File: docs/fundamentals/code-analysis/quality-rules/performance-warnings.md (hunk 5 above)
Comment (mpidash) [reply docs/fundamentals/code-analysis/quality-rules/performance-warnings.md]: Updated in [e055b1c](https://github.com/dotnet/docs/pull/36379/commits/e055b1cae46fe07e6c02527aff246dce1e4ada9c).

**Comment by gewarren** [comment]
@mpidash Looks like the rule ID needs to be updated to CA1868.

File: docs/fundamentals/code-analysis/quality-rules/ca1868.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+|                                     | Value                                  |
+| ----------------------------------- |----------------------------------------|
+| **Rule ID**                         | CA1868                                 |
+| **Category**                        | [Performance](performance-warnings.md) |
+| **Fix is breaking or non-breaking** | Non-breaking                           |
+| **Enabled by default in .NET 7**    | No                                     |
+
+## Cause
+
+An <xref:System.Collections.Generic.ISet%601.Add%2A> or <xref:System.Collections.Generic.ICollection%601.Remove%2A> (<xref:System.Collections.Immutable.IImmutableSet%601.Add%2A?displayProperty=nameWithType> or <xref:System.Collections.Immutable.IImmutableSet%601.Remove%2A?displayProperty=nameWithType>) call is guarded by a call to <xref:System.Collections.Generic.ICollection%601.Contains%2A> (<xref:System.Collections.Immutable.IImmutableSet%601.Contains%2A?displayProperty=nameWithType>).
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1868.md:28]:
```suggestion
An <xref:System.Collections.Generic.ISet%601.Add%2A?displayProperty=nameWithType> or <xref:System.Collections.Generic.ICollection%601.Remove%2A?displayProperty=nameWithType> call is guarded by a call to <xref:System.Collections.Generic.ICollection%601.Contains%2A>. Or, 
an <xref:System.Collections.Immutable.IImmutableSet%601.Add%2A?displayProperty=nameWithType> or <xref:System.Collections.Immutable.IImmutableSet%601.Remove%2A?displayProperty=nameWithType> call is guarded by a call to <xref:System.Collections.Immutable.IImmutableSet%601.Contains%2A?displayProperty=nameWithType>.
```

File: docs/fundamentals/code-analysis/quality-rules/ca1868.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+|                                     | Value                                  |
+| ----------------------------------- |----------------------------------------|
+| **Rule ID**                         | CA1868                                 |
+| **Category**                        | [Performance](performance-warnings.md) |
+| **Fix is breaking or non-breaking** | Non-breaking                           |
+| **Enabled by default in .NET 7**    | No                                     |
+
+## Cause
+
+An <xref:System.Collections.Generic.ISet%601.Add%2A?displayProperty=nameWithType> or <xref:System.Collections.Generic.ICollection%601.Remove%2A?displayProperty=nameWithType> call is guarded by a call to <xref:System.Collections.Generic.ICollection%601.Contains%2A>. Or,
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1868.md:28]:
```suggestion
An <xref:System.Collections.Generic.ISet%601.Add%2A?displayProperty=nameWithType> or <xref:System.Collections.Generic.ICollection%601.Remove%2A?displayProperty=nameWithType> call is guarded by a call to <xref:System.Collections.Generic.ICollection%601.Contains%2A>. Or, an <xref:System.Collections.Immutable.IImmutableSet%601.Add%2A?displayProperty=nameWithType> or <xref:System.Collections.Immutable.IImmutableSet%601.Remove%2A?displayProperty=nameWithType> call is guarded by a call to <xref:System.Collections.Immutable.IImmutableSet%601.Contains%2A?displayProperty=nameWithType>.
```

File: docs/fundamentals/code-analysis/quality-rules/ca1868.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+| ----------------------------------- |----------------------------------------|
+| **Rule ID**                         | CA1868                                 |
+| **Category**                        | [Performance](performance-warnings.md) |
+| **Fix is breaking or non-breaking** | Non-breaking                           |
+| **Enabled by default in .NET 7**    | No                                     |
+
+## Cause
+
+An <xref:System.Collections.Generic.ISet%601.Add%2A?displayProperty=nameWithType> or <xref:System.Collections.Generic.ICollection%601.Remove%2A?displayProperty=nameWithType> call is guarded by a call to <xref:System.Collections.Generic.ICollection%601.Contains%2A>. Or, 
+an <xref:System.Collections.Immutable.IImmutableSet%601.Add%2A?displayProperty=nameWithType> or <xref:System.Collections.Immutable.IImmutableSet%601.Remove%2A?displayProperty=nameWithType> call is guarded by a call to <xref:System.Collections.Immutable.IImmutableSet%601.Contains%2A?displayProperty=nameWithType>.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1868.md:29]:
```suggestion
```

**Review by gewarren — APPROVED** [review]

**Pull request merged**